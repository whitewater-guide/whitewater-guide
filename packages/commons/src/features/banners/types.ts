import { Connection, NamedNode, Node, NodeRef } from '../../apollo';
import { Group } from '../groups';
import { Region } from '../regions';

export enum BannerPlacement {
  MOBILE_SECTION_DESCRIPTION = 'mobileSectionDescription',
  MOBILE_SECTION_ROW = 'mobileSectionRow',
  MOBILE_SECTION_MEDIA = 'mobileSectionMedia',
  MOBILE_REGION_DESCRIPTION = 'mobileRegionDescription',
}

export enum BannerKind {
  Image = 'Image',
  WebView = 'WebView',
}

export interface BannerSource {
  kind: BannerKind;
  ratio: number | null;
  src: string;
}

export interface BannerCore extends NamedNode {
  enabled: boolean;
  slug: string;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
}

export interface Banner extends BannerCore {
  // --- connections
  regions?: Connection<Region>;
  groups?: Connection<Group>;
  // ---- transient
  deleted?: boolean;
}

export interface BannerSourceInput {
  kind: BannerKind;
  ratio: number | null;
  src: string | null;
}

export interface BannerInput {
  id: string | null;
  slug: string;
  name: string;
  enabled: boolean;
  priority: number;
  placement: BannerPlacement;
  source: BannerSourceInput;
  link: string | null;
  extras: { [key: string]: any } | null;
  // --- connections
  regions: NodeRef[];
  groups: NodeRef[];
}

export function isBanner(node: Node): node is Banner {
  return node.__typename === 'Banner';
}
