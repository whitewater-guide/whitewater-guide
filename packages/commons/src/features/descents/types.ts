import { Timestamped, PageInfo } from '../../apollo';
import { Section } from '../sections';
import { User } from '../users';

export interface DescentLevel {
  __typename?: 'Level';
  unit?: string | null;
  value?: number | null;
}

export interface DescentLevelInput {
  unit?: string | null;
  value: number;
}

export interface Descent extends Timestamped {
  __typename?: 'Descent';
  id: string;
  user: User;
  section: Section;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevel | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentEdge<TCursor = any> {
  __typename?: 'DescentEdge';
  node: Descent;
  cursor: TCursor;
}

export interface DescentInput {
  id?: string | null;

  sectionId: string;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevelInput | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentsConnection<TCursor = any> {
  __typename?: 'DescentsConnection';
  edges: DescentEdge[];
  pageInfo: PageInfo<TCursor>;
}

export interface DescentsFilter {
  startDate?: string | null;
  endDate?: string | null;
  difficulty?: [number, number] | null;
  userId?: string | null;
  sectionId?: string | null;
  sectionName?: string | null;
}
