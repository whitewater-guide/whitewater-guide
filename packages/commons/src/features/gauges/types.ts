import { NamedNode, Node, Timestamped } from '../../core';
import { HarvestStatus, Measurement } from '../measurements';
import { Point, PointInput } from '../points';
import { Source } from '../sources';

export interface Gauge extends NamedNode, Timestamped {
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
  enabled: boolean;
  location: Point;
  source: Source;
  // measurements?: Measurement[];
  lastMeasurement: Measurement | null;
  status: HarvestStatus | null;
}

export interface GaugeInput {
  id: string | null;
  name: string;
  source: Node;
  location: PointInput | null;
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
}
