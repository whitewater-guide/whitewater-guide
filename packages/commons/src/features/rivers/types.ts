import {
  Connection,
  NamedNode,
  Node,
  NodeRef,
  Timestamped,
} from '../../apollo';
import { Region } from '../regions';
import { Section } from '../sections';

export interface River extends NamedNode, Timestamped {
  region: Region;
  sections: Connection<Section>;
  altNames: string[];
}

export interface RiverInput {
  id: string | null;
  name: string;
  altNames: string[] | null;
  region: NodeRef;
}

export interface RiversFilter {
  search?: string;
  regionId?: string;
}

export const isRiver = (node?: Node | null): node is River =>
  !!node && node.__typename === 'River';
