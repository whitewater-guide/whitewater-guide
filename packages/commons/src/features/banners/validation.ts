import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';

import { BannerKind, BannerPlacement } from './types';

const BannerSourceSchema = yup
  .object({
    kind: yup.mixed().oneOf(Object.values(BannerKind)),
    url: yupTypes.nonEmptyString(),
  })
  .noUnknown();

export const BannerInputSchema = yup
  .object({
    id: yupTypes.uuid().defined().nullable(),
    slug: yupTypes.slug(),
    name: yupTypes.nonEmptyString(),
    priority: yup.number().integer(),
    enabled: yup.bool(),
    placement: yup.mixed().oneOf(Object.values(BannerPlacement)),
    source: BannerSourceSchema.clone(),
    link: yupTypes.https().nullable(true),
    extras: yup.object().nullable(true),
    regions: yup.array(yupTypes.node()),
    groups: yup.array(yupTypes.node()),
  })
  .strict(true)
  .noUnknown(true);
