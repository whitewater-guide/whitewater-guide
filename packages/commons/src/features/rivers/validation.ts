// tslint:disable-next-line
import * as Joi from 'joi';

export const RiverInputSchema = Joi.object().keys({
  id: Joi.string().guid().allow(null),
  name: Joi.string().min(2).max(200),
  region: Joi.object().keys({ id: Joi.string().guid() }),
  altNames: Joi.array().items(Joi.string().min(2)).allow(null),
});
