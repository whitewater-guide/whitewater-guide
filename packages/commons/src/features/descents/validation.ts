import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';
import { DescentLevelInput, DescentInput } from './types';

export const DescentLevelInputSchema = yup
  .object<DescentLevelInput>({
    unit: yup
      .string()
      .max(20)
      .notRequired()
      .nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema = yup
  .object<DescentInput>({
    id: yupTypes
      .uuid()
      .notRequired()
      .nullable(),
    sectionId: yupTypes
      .uuid()
      .required()
      .nullable(false),
    startedAt: yup.date().required() as any,
    duration: yup
      .number()
      .integer()
      .notRequired()
      .nullable(),
    level: DescentLevelInputSchema.clone()
      .notRequired()
      .nullable(),
    comment: yup
      .string()
      .notRequired()
      .nullable(),
    public: yup
      .bool()
      .notRequired()
      .nullable(),
  })
  .strict(true)
  .noUnknown();
