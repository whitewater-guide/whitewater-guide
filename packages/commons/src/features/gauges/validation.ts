import { baseStruct } from '../../utils/validation';
import { PointInputStruct } from '../points';

const GaugeInputFields = {
  id: baseStruct.union(['uuid', 'null']),
  name: 'nonEmptyString',
  code: 'nonEmptyVarchar',
  levelUnit: 'nonEmptyVarchar|null',
  flowUnit: 'nonEmptyVarchar|null',
  location: baseStruct.union([PointInputStruct, 'null']),
  requestParams: 'object|null',
  cron: baseStruct.union(['cron', 'null']),
  url: baseStruct.union(['url', 'null', baseStruct.literal('')]),
  source: 'node',
};

export const GaugeInputStruct = baseStruct.object(GaugeInputFields);

export const GaugeFormStruct = baseStruct.object({
  ...GaugeInputFields,
  requestParams: baseStruct.union(['jsonString', 'null']),
});
