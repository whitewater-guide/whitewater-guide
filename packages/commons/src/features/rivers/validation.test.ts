import { RiverInput } from './types';
import { RiverInputSchema } from './validation';
import { createSafeValidator } from '@whitewater-guide/validation';

const validator = createSafeValidator(RiverInputSchema);

type TestValue = [string, RiverInput];

const correct: RiverInput = {
  id: '7ce3319e-c264-11e8-a355-529269fb1459',
  name: 'foo',
  region: { id: '879ca106-c264-11e8-a355-529269fb1459' },
  altNames: ['bar', 'baz'],
};

const correctValues: TestValue[] = [
  ['full value', correct],
  [
    'null value',
    {
      ...correct,
      id: null,
      altNames: null,
      importId: null,
    },
  ],
  [
    'with import id',
    {
      ...correct,
      importId: 'green',
    },
  ],
];

const incorrectValues: TestValue[] = [
  ['bad uuid', { ...correct, id: 'foo' }],
  ['empty name', { ...correct, name: '' }],
  ['empty alt names', { ...correct, altNames: [''] }],
  ['bad region', { ...correct, region: { id: 'foo' } }],
  ['extra fields', { ...correct, foo: 'bar' } as any],
];

it.each(correctValues)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrectValues)('should be invalid for %s', (_, value) => {
  const error = validator(value);
  expect(error).not.toBeNull();
  expect(error).toMatchSnapshot();
});
