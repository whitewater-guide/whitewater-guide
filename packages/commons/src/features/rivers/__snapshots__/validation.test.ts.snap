// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for bad region 1`] = `
Object {
  "region": Object {
    "id": "yup:string.uuid",
  },
}
`;

exports[`should be invalid for bad uuid 1`] = `
Object {
  "id": "yup:string.uuid",
}
`;

exports[`should be invalid for empty alt names 1`] = `
Object {
  "altNames": Array [
    "yup:string.nonEmpty",
  ],
}
`;

exports[`should be invalid for empty name 1`] = `
Object {
  "name": "yup:string.nonEmpty",
}
`;

exports[`should be invalid for extra fields 1`] = `
Object {
  "undefined": Object {
    "key": "yup:object.noUnknown",
    "options": Object {
      "path": "this",
    },
  },
}
`;
